zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentReconstructionStarter.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.zeebe.engine.processing.deployment;
+
+import io.camunda.zeebe.engine.state.immutable.DeploymentState;
+import io.camunda.zeebe.protocol.impl.record.value.deployment.DeploymentRecord;
+import io.camunda.zeebe.protocol.record.intent.DeploymentIntent;
+import io.camunda.zeebe.stream.api.ReadonlyStreamProcessorContext;
+import io.camunda.zeebe.stream.api.StreamProcessorLifecycleAware;
+import io.camunda.zeebe.stream.api.scheduling.Task;
+import io.camunda.zeebe.stream.api.scheduling.TaskResult;
+import io.camunda.zeebe.stream.api.scheduling.TaskResultBuilder;
+import java.time.Duration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class DeploymentReconstructionStarter implements StreamProcessorLifecycleAware, Task {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DeploymentReconstructionStarter.class);
+  private final DeploymentState deploymentState;
+
+  public DeploymentReconstructionStarter(final DeploymentState deploymentState) {
+    this.deploymentState = deploymentState;
+  }
+
+  @Override
+  public void onRecovered(final ReadonlyStreamProcessorContext context) {
+    if (deploymentState.hasStoredAllDeployments()) {
+      LOG.trace("All deployments are already stored, skipping reconstruction");
+      return;
+    }
+
+    context.getScheduleService().runDelayed(Duration.ZERO, this);
+  }
+
+  @Override
+  public TaskResult execute(final TaskResultBuilder taskResultBuilder) {
+    LOG.debug("Not all deployments are stored, starting reconstruction");
+    taskResultBuilder.appendCommandRecord(DeploymentIntent.RECONSTRUCT, new DeploymentRecord());
+    return taskResultBuilder.build();
+  }
```

Comment:
❓ So what is the plan here? Will you write all the reconstruct commands in this starter? That seems like a bad idea as if there's a lot of processes it could fill up the log significantly.

On the other hand you only have a `DeploymentRecord`. How will you use this to gradually reconstruct the deployment state?

Reply:
My plan was to do it incrementally, i.e. write the first `Reconstruct` command which writes the next one etc. That way we automatically benefit from batch processing too.
On `Deployment/Reconstructed` we can update the state based on the contained resources in the `DeploymentRecord`.

Reply:
So basically:
- Get the first resource key (whether it's a process/form/drd)
- Write the resoncstruct command for this key
- In that processor get the next resource key (whether it's a process/form/drd)
- Write a followup reconstruct command for this key

I guess you know you're done once you find a resource that contains a deployment key that isn't -1?

Reply:
> I guess you know you're done once you find a resource that contains a deployment key that isn't -1?

No, there can also be resources with a deployment key but that deployment key doesn't exist.

I think the sequence should be this: 
1. Write a Reconstruct command (no key necessary)
2. On processing, look for the next resource.
3. Find all resources with the same deployment key if it has one
4. Write a Reconstructed event with the rebuilt DeploymentRecord
5. Write the next Reconstruct command
6. If we can't find a resource without a matching deployment write the ReconstructedAll event and no Reconstruct command.